pkg/kubectl/resource_columns.go

```diff
@@ -0,0 +1,551 @@
+
+		if content.Kind() != reflect.Func {
+			err = fmt.Errorf("invalid print handler. %#v is not a function", entry.content)
+			return
+		}
+		funcType := content.Type()
+
+		args := []reflect.Value{}
+		for i := 0; i < funcType.NumIn(); i++ {
+			if funcType.In(i) == objType {
+				args = append(args, vof(obj))
+				continue
+			}
+			if funcType.In(i) == reflect.TypeOf((*runtime.Object)(nil)).Elem() {
+				args = append(args, vof(obj))
+				continue
+			}
+			if funcType.In(i) == contextType {
+				args = append(args, vof(c))
+				continue
+			}
+			err = fmt.Errorf("Cannot populate type %v for %s", funcType.In(i), entry.label)
+			return
+		}
+		resultValue := content.Call(args)
+
+		if funcType.Out(0) == reflect.TypeOf((*string)(nil)).Elem() {
+			result = append(result, []string{resultValue[0].String()})
+		} else if funcType.Out(0) == reflect.TypeOf((*[]string)(nil)).Elem() {
+			allRows := []string{}
+			for i := 0; i < resultValue[0].Len(); i++ {
+				allRows = append(allRows, resultValue[0].Index(i).String())
+			}
+			result = append(result, allRows)
+		} else if funcType.Out(0) == reflect.TypeOf((*error)(nil)).Elem() {
+			if !resultValue[0].IsNil() {
+				err, _ = resultValue[0].Interface().(error)
+				return
+			}
+		} else {
+			err = fmt.Errorf("Wrong return type %v for %s", funcType.In(0), entry.label)
+			return
+		}
+
+		if funcType.NumOut() == 2 {
+			if funcType.Out(1) == reflect.TypeOf((*error)(nil)).Elem() {
+				if !resultValue[1].IsNil() {
+					err, _ = resultValue[1].Interface().(error)
+					return
+				}
+			} else {
+				err = fmt.Errorf("Wrong return type %v for %s", funcType.In(1), entry.label)
+				return
+			}
+		}
+	}
+	return
+}
+
+var columnMap = map[reflect.Type][]columnDefinition{
+	reflect.TypeOf((*api.Pod)(nil)):                   podColumnsFunc,
+	reflect.TypeOf((*api.PodTemplate)(nil)):           podTemplateColumnsFunc,
+	reflect.TypeOf((*api.ReplicationController)(nil)): replicationControllerColumnsFunc,
+	reflect.TypeOf((*api.Service)(nil)):               serviceColumnsFunc,
+	reflect.TypeOf((*api.Endpoints)(nil)):             endpointsColumnsFunc,
+	reflect.TypeOf((*api.Node)(nil)):                  nodeColumnsFunc,
+	reflect.TypeOf((*api.Event)(nil)):                 eventColumnsFunc,
+	reflect.TypeOf((*api.LimitRange)(nil)):            limitRangeColumnsFunc,
+	reflect.TypeOf((*api.ResourceQuota)(nil)):         resourceQuotaColumnsFunc,
+	reflect.TypeOf((*api.Namespace)(nil)):             namespaceColumnsFunc,
+	reflect.TypeOf((*api.Secret)(nil)):                secretColumnsFunc,
+	reflect.TypeOf((*api.ServiceAccount)(nil)):        serviceAccountColumnsFunc,
+	reflect.TypeOf((*api.PersistentVolume)(nil)):      persistentVolumeColumnsFunc,
+	reflect.TypeOf((*api.PersistentVolumeClaim)(nil)): persistentVolumeClaimColumnsFunc,
+	reflect.TypeOf((*api.ComponentStatus)(nil)):       componentStatus,
+}
+
+func getColumnsFor(obj runtime.Object, c printContext) (columns []columnDefinition, err error) {
```

Comment:
Yes, you will get some type safety but at the cost of bloating the implementation - 5-6 lines of code for a single column instead of one. Plus you will have to register all types that an api object should be casted to in the order the columns are printed.

And when you register the interface you could also provide the column label instead defining a function to do so. Then the interface will have only one method - content. So we will be creating a new type with a single function that takes no parameter:

```go
type SASecrets *api.ServiceAccount
func (p SASecrets) Content() string { return strings.Join(p.Secrets, "," ) }

(...)
secretColumns.register("SECRETS", reflect.TypeOf((*SASecrets)(nil)))
```
And now, why do we need the extra type at all? 

```go
func SASecretContent(p *api.ServiceAccount) string { return strings.Join(p.Secrets, "," ) }
(...)
secretColumns.register("SECRETS", SASecretContent)
```
And here we are more or less where this design is right now.